### PR TITLE
Feat/new badge creation

### PR DESCRIPTION
**Pull request by jona159**
This PR introduces the new badge creation process, which stores the competencies of a badge in an extension. The category of a badge is also stored in an extension. A new design was integrated, but it is not quite finished (on mycelia`s side). Old extensions are still also stored for now, because removing them breaks things.

### CODE REVIEW

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 1]
```diff
@@ -73,7 +73,7 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
 		return badges && badges.length;
 	}
 	readonly issuerImagePlacholderUrl = preloadImageURL(
-		'../../../../breakdown/static/images/placeholderavatar-issuer.svg'
+		'../../../../breakdown/static/images/placeholderavatar-issuer.svg',
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:76]: Same as above

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 2]
```diff
@@ -120,28 +118,28 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
...
+		private externalToolsManager: ExternalToolsManager,
 	) {
 		super(router, route, sessionService);
 
 		this.badgeClassLoaded = badgeManager.badgeByIssuerSlugAndSlug(this.issuerSlug, this.badgeSlug).then(
 			(badge) => {
 				this.badgeClass = badge;
 				this.title.setTitle(
-					`Badge Class - ${this.badgeClass.name} - ${this.configService.theme['serviceName'] || 'Badgr'}`
+					`Badge Class - ${this.badgeClass.name} - ${this.configService.theme['serviceName'] || 'Badgr'}`,
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:129]: Same as above

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 3]
```diff
@@ -120,28 +118,28 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
...
 				);
 				this.loadInstances();
 			},
 			(error) =>
 				this.messageService.reportLoadingError(
 					`Cannot find badge ${this.issuerSlug} / ${this.badgeSlug}`,
-					error
-				)
+					error,
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:136]: Same as above

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 4]
```diff
@@ -120,28 +118,28 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
...
+				),
 		);
 
 		this.issuerLoaded = issuerManager.issuerBySlug(this.issuerSlug).then(
 			(issuer) => (this.issuer = issuer),
-			(error) => this.messageService.reportLoadingError(`Cannot find issuer ${this.issuerSlug}`, error)
+			(error) => this.messageService.reportLoadingError(`Cannot find issuer ${this.issuerSlug}`, error),
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:142]: Same as above

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 5]
```diff
@@ -171,10 +169,10 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
 			},
 			(error) => {
 				this.messageService.reportLoadingError(
-					`Could not load recipients ${this.issuerSlug} / ${this.badgeSlug}`
+					`Could not load recipients ${this.issuerSlug} / ${this.badgeSlug}`,
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:172]: Same as above

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 6]
```diff
@@ -200,11 +198,11 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
 						},
 						(error) =>
 							this.messageService.reportAndThrowError(
-								`Failed to revoke badge to ${instance.recipientIdentifier}`
-							)
+								`Failed to revoke badge to ${instance.recipientIdentifier}`,
+							),
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:202]: Same as above

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 7]
```diff
@@ -200,11 +198,11 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
...
 					);
 				},
-				() => void 0 // Cancel
+				() => void 0, // Cancel
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:205]: Same as above

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 8]
```diff
@@ -226,12 +224,12 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
 							},
 							(error) => {
 								this.messageService.reportAndThrowError(
-									`Failed to delete badge class: ${BadgrApiFailure.from(error).firstMessage}`
+									`Failed to delete badge class: ${BadgrApiFailure.from(error).firstMessage}`,
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:227]: And so on, I won't add a comment to every single occurrence. But why did you add commas everywhere?

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html [hunk 9]
```diff
@@ -62,55 +77,377 @@ <h3 class="u-text-body-bold">Badge Basics</h3>
...
+
+					<ng-template #itemTemplate let-item>
+						<a [innerHTML]="item.name"></a>
+					</ng-template>
+
+					<ng-template #notFoundTemplate let-notFound>
+						<div>Dieser Tag existiert noch nicht.</div>
+					</ng-template>
 				</div>
+				<bg-formfield-select
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:142]: Clarify if the autocomplete *and* select is necessary

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html [hunk 10]
```diff
@@ -62,55 +77,377 @@ <h3 class="u-text-body-bold">Badge Basics</h3>
...
+				<button style="margin-top: 2em; background-color: #492e98" class="button">
+					<a
+						style="color: white"
+						href="https://esco.ec.europa.eu/en/classification/skill_main"
+						target="_blank"
+						>Bring mich zu ESCO</a
+					>
+				</button>
+
+				<!-- <ng-container *ngIf="badgeClassForm.controls.competencies.controls.length === 0; else competencyLoop">
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:287]: What is this commented code doing here?

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html [hunk 11]
```diff
@@ -62,55 +77,377 @@ <h3 class="u-text-body-bold">Badge Basics</h3>
...
+							<button
+								*ngIf="badgeClassForm.controls.competencies.controls.length > 0"
+								type="button"
+								(click)="removeCompetency(competency)"
+								class="u-text-link"
+							>
+								Remove
+							</button>
+						</div>
+						<!-- <button
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:433]: Commented code

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html [hunk 12]
```diff
@@ -62,55 +77,377 @@ <h3 class="u-text-body-bold">Badge Basics</h3>
...
+							style="background-color: var(--color-grün); color: var(--color-blue); margin-top: 2em"
+						>
+							Hinzufügen
+						</button> -->
+					</div>
+				</ng-container>
+				<!-- </ng-template> -->
+				<div class="l-stack l-stack-right u-margin-top2x">
+					<button (click)="addCompetency()" type="button" class="button button-secondary">
+						Kompetenz hinzufügen
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:444]: There seem to be two "Hinzufügen" buttons

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html [hunk 13]
```diff
@@ -161,7 +498,7 @@ <h3 class="u-text-body-bold">Was sind die Kriterien?</h3>
 	</div>
 
 	<!-- Criteria Panel -->
-	<div class="formsection">
+	<!-- <div class="formsection">
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:503]: Commented code

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html [hunk 14]
```diff
@@ -346,7 +683,7 @@ <h3 class="u-text-body-bold u-text-dark1">
 	</div>
 
 	<!-- tags -->
-	<div *ngIf="tagsEnabled" class="formsection">
+	<!-- <div *ngIf="tagsEnabled" class="formsection">
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:688]: Commented code

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html [hunk 15]
```diff
@@ -515,14 +837,14 @@ <h2 class="formsection-x-title">Optionale Details</h2>
 		</div>
 	</div>
 
-	<div class="formsection formsection-nohelp">
+	<!-- <div class="formsection formsection-nohelp">
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:842]: Commented code

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts [hunk 16]
```diff
@@ -380,6 +392,15 @@ export class BadgeClassEditFormComponent extends BaseAuthenticatedRoutableCompon
 					id: index,
 					name: tag,
 				}));
+				outerThis.tagOptions = outerThis.existingTags.map(
+					(tag) =>
+						({
+							//@ts-ignore
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:398]: Why is this ignore comment necessary?

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts [hunk 17]
```diff
@@ -508,6 +512,32 @@ export class BadgeClassEditFormComponent extends BaseAuthenticatedRoutableCompon
...
+					dialogBody: 'Are you sure you want to remove this competency? This action cannot be undone.',
+					resolveButtonLabel: 'Remove Competency',
+					rejectButtonLabel: 'Cancel',
+				}))
+			) {
+				return;
+			}
+		}
+
+		this.badgeClassForm.controls.competencies.removeAt(
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:536]: It seems that this should be inside the body of the previous if, because otherwise this will try to remove an empty competence

**Review by timber-they — CHANGES_REQUESTED** [review]

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts (hunk 16 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts]: removed it and used a better type for existingTags

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html (hunk 10 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:287]: There is a comment above it now, explaining it

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts [hunk 18]
```diff
@@ -60,7 +60,7 @@ export class BadgeClassDetailComponent extends BaseAuthenticatedRoutableComponen
 
 	get activeRecipientCount() {
 		const badges = this.allBadgeInstances.entities.filter(
-			(thisEntity) => !thisEntity.isExpired && !thisEntity.isRevoked
+			(thisEntity) => !thisEntity.isExpired && !thisEntity.isRevoked,
```
[the remark this answers is not in the document]
Comment (jona159) [reply src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:63]: Sorry, prettier added the comments automatically. Do you think i should remove them?

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html (hunk 11 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html]: Removed

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html (hunk 13 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:497]: Removed

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html (hunk 14 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html]: Removed

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html (hunk 15 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html]: Removed

File: src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (timber-they) [reply src/app/issuer/components/badgeclass-detail/badgeclass-detail.component.ts:63]: Hm it seems unnecessary, but if it's supposed to be prettier maybe there's a reason for it. So I guess it can stay like that

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html (hunk 10 above)
Comment (timber-they) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:287]: Hm but does this TODO have its own ticket then?

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts (hunk 17 above)
Comment (timber-they) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:534]: Ah I see. But then the if structure is a bit weird; like, why are there two ifs?

Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:561]: Tbh i basically copied the ```removeAlignment``` function, i am not sure why this structure was chosen. I would suggest to leave it as is for now, but to combine the functions once we are certain that the badge creation process is finished. What do you think, should i add a _TODO_ and ticket for that?

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html (hunk 12 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:444]: I commented one out for now, will clarify which one we should use

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html (hunk 10 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.html:290]: Now it [does](https://github.com/orgs/mint-o-badges/projects/2/views/1?pane=issue&itemId=57524205)

**Comment by jona159** [comment]
Dont merge yet

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts (hunk 17 above)
Comment (timber-they) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:561]: If it's the same then I think it is alright

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts [hunk 19]
```diff
@@ -371,6 +371,33 @@ export class BadgeClassEditFormComponent extends BaseAuthenticatedRoutableCompon
...
+					dialogTitle: 'Wenn du die Kategorie änderst, werden alle Kompetenzen gelöscht.',
+					dialogBody: 'Möchtest du fortfahren?',
+					resolveButtonLabel: 'Fortfahren',
+					rejectButtonLabel: 'Abbrechen',
+				})
+			) {
+				this.badgeClassForm.controls.competencies.reset();
+				const controls = this.badgeClassForm.controls.competencies.controls;
+				for (let i = controls.length - 1; i >= 0; i--) {
+					this.badgeClassForm.controls.competencies.removeAt(controls.indexOf(controls[i]));
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:390]: Why is this so complicated? Like, why do you delete the competencies at the index of the controls of the competencies? Can't you just directly delete all competencies, e.g. with a `clear` call or sth similar? And if not, why don't iterate over the competencies instead of the controls of the competencies?

**Review by timber-they — CHANGES_REQUESTED** [review]

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts (hunk 19 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts]: My bad, iterating over competencies now

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts [hunk 20]
```diff
@@ -384,10 +374,10 @@ export class BadgeClassEditFormComponent extends BaseAuthenticatedRoutableCompon
...
 				})
 			) {
-				this.badgeClassForm.controls.competencies.reset();
-				const controls = this.badgeClassForm.controls.competencies.controls;
-				for (let i = controls.length - 1; i >= 0; i--) {
-					this.badgeClassForm.controls.competencies.removeAt(controls.indexOf(controls[i]));
+				const competencies = this.badgeClassForm.controls.competencies;
+				competencies.reset();
+				for (let i = competencies.length - 1; i >= 0; i--) {
+					this.badgeClassForm.controls.competencies.removeAt(competencies[i]);
```
Comment (timber-they) [inline src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts:380]: I still don't understand, why you have to remove the entries one by one. If this is in fact necessary, maybe a comment explaining why would be nice. If not, surely there exists a method that removes them all (like `clear`)?
Even if not, I don't think this would work that way, because now you're calling `removeAt` with the competency, instead of the index. I think what should work would be:
```ts
for (let i = competencies.length - 1; i >= 0; i--) {
    competencies.removeAt(i);
}
```

**Review by timber-they — CHANGES_REQUESTED** [review]

File: src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts (hunk 20 above)
Comment (jona159) [reply src/app/issuer/components/badgeclass-edit-form/badgeclass-edit-form.component.ts]: Did not find a method that would remove all at once, i guess the closes thing would be to call ```competencies.setValue([])```. Not sure if thats better than removing the competencies, for now i pushed the fix you suggested

**Review by timber-they — APPROVED** [review]
